React Part 2: GTSFM Graph and Point Cloud Components

adisingh50 opened the pull request:
This is the second PR related to integrating my React application into the gtsfm repository. The components I've added are used to render the primary GTSFM landing page graph as well as the point cloud generated after data association. 

- `Components` folder stores modular components (writted in JS)
- `stylesheets` folder contains stylings (written in CSS) for various JS components

With this updated react code, this is what can be rendered on screen:
![image](https://user-images.githubusercontent.com/40643713/113486851-bd72aa80-9482-11eb-9b64-f6d455409468.png)

![image](https://user-images.githubusercontent.com/40643713/113486866-d0857a80-9482-11eb-9241-591be602606c.png)

adisingh50 commented on rtf_vis_tool/package.json:
```diff
@@ -14,7 +14,7 @@
     "react-three-fiber": "^5.3.12",
     "react-xarrows": "^1.5.2",
-    "three": "^0.127.0"
+    "three": "^0.124.0"
```
small dependency version conflict here. Just had to downgrade three to fix it

adisingh50 replied:
and because I modified the version of the `three` dependency within `package.json`, that also modified the `package-lock.json` file as well (no noteworthy changes though)

adisingh50 commented on rtf_vis_tool/src/App.js:
```diff
@@ -3,13 +3,13 @@ import {extend} from "react-three-fiber";
+    <DivGraph/>
   )
 }
 
```
updated App.js (which was included in React PR #1) to now render DivGraph.js as the landing page

adisingh50 commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -0,0 +1,68 @@
+    )
+}
+
+export default CoordinateGrid;
```
A Helper component used in Point Cloud rendering to display the x, y, and z axes. Also renders some grid lines along the xy plane

adisingh50 commented on rtf_vis_tool/src/Components/Data_Association_PC.js:
```diff
@@ -0,0 +1,113 @@
+    )
+}
+
+export default Data_Association_PC;
```
The component which actually utilizes React Three Fiber to render the point cloud.

adisingh50 commented on rtf_vis_tool/src/Components/DivGraph.js:
```diff
@@ -0,0 +1,164 @@
+    )
+}
+
+export default DivGraph;
```
DivGraph.js serves as the landing page graph. Renders 43 nodes, the edges connecting them, and the 5 plates.

adisingh50 commented on rtf_vis_tool/src/Components/DivNode.js:
```diff
@@ -0,0 +1,37 @@
+    )
+}
+
+export default DivNode;
```
a basic node rendered in the graph (with styling included)

adisingh50 replied:
when clicked, it just alerts some text on the screen (which is a temporary/placeholder functionality)

adisingh50 commented on rtf_vis_tool/src/Components/OrbitControlsComponent.js:
```diff
@@ -0,0 +1,23 @@
+    )
+}
+
+export default OrbitControlsComponent;
```
This component enables the user to pan and orbit around a point cloud.

adisingh50 commented on rtf_vis_tool/src/Components/PointMesh.js:
```diff
@@ -0,0 +1,14 @@
+    )
+}
+
+export default PointMesh;
```
a component for a single spherical point in a point cloud

adisingh50 commented on rtf_vis_tool/src/Components/SfMDataDivNode.js:
```diff
@@ -0,0 +1,35 @@
+    )
+}
+
+export default SfMDataDivNode;
```
like the DivNode.js component, but is specifically used to render the SfM Node since it has custom functionality when clicked

adisingh50 commented on rtf_vis_tool/src/Components/gtsfm_edge_list.js:
```diff
@@ -0,0 +1,58 @@
+                        ['Dense Mesh Reconstruction', 'Aggregate'], 
+                        ['Aggregate', 'Zipped Results for All Scenes']]
+
+module.exports = gtsfm_edge_list;
```
raw list of all the edges in the GTSFM graph

adisingh50 commented on rtf_vis_tool/src/Components/gtsfm_node_list.js:
```diff
@@ -0,0 +1,53 @@
+    node_set_formatted.push({id: gtsfm_node_set[i]});
+}
+
+module.exports = node_set_formatted;
```
raw list of all the GTSFM nodes in the graph

adisingh50 commented on rtf_vis_tool/src/stylesheets/Data_Association_PC.css:
```diff
@@ -0,0 +1,45 @@
+    top: 15%;
+    cursor: pointer;
+    font-size: large;
+}
```
css stylings for the Data_Assocation_PC.js component

adisingh50 commented on rtf_vis_tool/src/stylesheets/DivGraph.css:
```diff
@@ -0,0 +1,118 @@
+    border: 2px solid #0b8db5;
+    padding-left: 1%;   
+    z-index: -1;
+}
```
css stylings for the DivGraph.js component

adisingh50 commented on rtf_vis_tool/src/Components/DivGraph.js:
```diff
@@ -0,0 +1,164 @@
+            {/* {showFS && <FrontendSummary json={fs_json} toggleFS={toggleFrontEndSummaryDisplay}/>} */}
+            {/* {showMVO && <MVOSummary json={mvo_json} toggleMVO={toggleMVOMetrics}/>} */}
+            {showDA_PC && <Data_Association_PC da_json={da_json} rotated_json={rotated_da_json} toggleDA_PC={toggleDataAssoc_PointCloud}/>}
+            {/* {showRSS && <RotSuccessSummary json={fs_json} toggleRSS={toggleRotSummaryDisplay}/>} */}
```
some of this is commented out because they are attached to components which I have not yet added to this PR (for simplicity's sake). In the 3rd PR, I will add the components related to these and appropriately uncomment these lines.

johnwlambert commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -0,0 +1,68 @@
+    return (
+      <mesh>
+          <Line 
+            points={[[-10,0,0], [10,0,0]]} 
```
10 and 50 seem like magic numbers -- is there a simple way to avoid that?

adisingh50 replied:
this was just a stylistic choice to make it render nicely. I can wrap these in variables and document them so it doesn't seem hard coded.

johnwlambert commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -0,0 +1,68 @@
+}
+
+//Component which renders the x,y,z axes along with the grid lines along xy plane
+const CoordinateGrid = () => { 
```
could do the docstring be more detailed -- are we returning html here? seems nonstandard for js?

ayushbaid replied:
do you think this can go into a config? that will solve the magic numbers issue?

adisingh50 replied:
I have added a more detailed docstring about the return type. Specifically, every single JS file inside the `Components` directory returns the format JSX, which is essentially a modified version of HTML that is meant to be readable by React code. So while it looks like HTML, it's technically not HTML. It's a syntax that React requires to render it's components. I can probably add this description into the README.md file, since JSX format applies to every file I have in the `Components` folder.

johnwlambert commented on rtf_vis_tool/src/Components/Data_Association_PC.js:
```diff
@@ -0,0 +1,113 @@
+    const [isCentered, setIsCentered] = useState(false);
+    const [showCoordGrid, setShowCoordGrid] = useState(true);
+    const lightgray = `rgb(163, 168, 165)`;
+    const pointSizeArr = [0.2,8,8];
```
what do 0.2 and 8 represent?

adisingh50 replied:
0.2 represents the radius of the points and 8s represent something called height and width segments. The 8s are actually optional parameters, so I have just left the 0.2 to keep it simple. Added a comment in the file as well about it.

johnwlambert commented on rtf_vis_tool/src/Components/Data_Association_PC.js:
```diff
@@ -0,0 +1,113 @@
+    const lightgray = `rgb(163, 168, 165)`;
+    const pointSizeArr = [0.2,8,8];
+
+    //render points_3d from data association json file (passed in through props)
```
can we switch to reading from the COLMAP points3D.txt file instead?

adisingh50 replied:
yup. I see you merged your `start-colmap-export` branch into master. I'll pull that onto my branch, run `scene_optimizer.py`, and then try rendering the point cloud from bundle adjustment output

johnwlambert commented on rtf_vis_tool/src/Components/DivGraph.js:
```diff
@@ -0,0 +1,164 @@
+import multiview_optimizer_json from '.././result_metrics/multiview_optimizer_metrics.json';
+
+//Landing Page Component to display GTSFM graph (what user first sees)
+const DivGraph = (props) => {
```
is there any better name for this?

adisingh50 replied:
sure, instead of `DivGraph.js`, I could call it `LandingPageGraph.js`, since that's more telling of the component's purpose

johnwlambert commented on rtf_vis_tool/src/Components/gtsfm_node_list.js:
```diff
@@ -0,0 +1,53 @@
+//List of all 43 Nodes in GTSFM Visualization Graph
+//imported in DivGraph.js
+
+const gtsfm_node_set = ['absolute Rs',
```
can we list these alphabetically, or in some topological order-based grouping for now?

adisingh50 replied:
yup, I'll order it alphabetically.

johnwlambert commented on rtf_vis_tool/src/stylesheets/Data_Association_PC.css:
```diff
@@ -0,0 +1,45 @@
+/* Stylings for Data_Association_PC.js Component */
+
+.da-container {
```
could we add comments explaining the choices for each of these css classes?

adisingh50 replied:
done

johnwlambert commented on rtf_vis_tool/src/stylesheets/DivGraph.css:
```diff
@@ -0,0 +1,118 @@
+/* Stylings for DivGraph.js Component */
+
+.div_graph_container {
```
same here about comments

adisingh50 replied:
done

ayushbaid reviewed (commented):
Nice work @adisingh50.

Does the web application run in the current state of the PR?

I get this error: ```Module not found: Can't resolve '.././result_metrics/data_association_metrics.json' in '/home/ayush/dev/gtsfm/rtf_vis_tool/src/Components'```

ayushbaid commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -0,0 +1,68 @@
+            color="blue" 
+            position={[0,0,0]}
+            lineWidth={0.3}
+          />
```
is it a standard practise to have one property per line?

adisingh50 replied:
Yup. For components that are like `<Html> </Html>`, developers usually write the properties on the same line, since extra code will come in between the tags. But for components that are like `<Line />`, developers usually write the properties on different lines for readability sake (since there is no extra code)

ayushbaid commented on rtf_vis_tool/src/Components/DivGraph.js:
```diff
@@ -0,0 +1,164 @@
+            {/* {showRSS && <RotSuccessSummary json={fs_json} toggleRSS={toggleRotSummaryDisplay}/>} */}
+
+            <div className="gtsfm_graph">
+                {/* Render 43 Graph Nodes */}
```
can we somehow templetize this? I fear this structure might lead to inconsistencies on updates.

adisingh50 replied:
how specifically would we templatize this? I too have thought about how clunky 43 lines of `<DivNode />` components are, but the issue is that I need to specify strictly unique `topOffset` and `leftOffset` properties for each of the 43 DivNodes. Which basically means I can't just do a for-loop through the node names and render them all concisely (which I would have liked to do in the first place)

ayushbaid commented on rtf_vis_tool/src/Components/DivGraph.js:
```diff
@@ -0,0 +1,164 @@
+
+                {/* Render Plates */}
+                <div className="scene_optimizer_plate">
+                    <p style={{color: 'red', fontWeight: 'bold'}}>Scene Optimizer Scenes</p>
```
plate style in css?

adisingh50 replied:
you're right. I should add this plate title styling in the css file

ayushbaid commented on rtf_vis_tool/src/Components/DivNode.js:
```diff
@@ -0,0 +1,37 @@
+const DivNode = (props) => {
+
+    // a standard styling for each node in the graph
+    const divNodeStyle = {
```
can this go in a css?

adisingh50 replied:
The only reason why this style cannot go in its own CSS is because of 3 styling properties on lines 9,10, and 11. These `top`, `left`, and `backgroundColor` stylings are unique in that they depend on `props`, which are passed into the component. These props can only be understood by JS code, not CSS code. Therefore, I need to keep this block of styling in JS.

ayushbaid commented on rtf_vis_tool/src/Components/SfMDataDivNode.js:
```diff
@@ -0,0 +1,35 @@
+const SfMDataDivNode = (props) => {
+
+    //similar styling applies to DivNode.js
+    const divNodeStyle = {
```
can we import the same variable/style? If two things have to be the same, I'd suggest storing just a single copy.

adisingh50 replied:
created the CSS file to store most of the duplicate stylings.

ayushbaid reviewed (commented):
I am not well-versed in react enough to review this, but left a few comments. Let me know if those don't make a lot of sense.

adisingh50 commented:
Thanks John and Ayush for the helpful comments. And Ayush, for that error you encountered, the reason for it is because I still need to add a new PR that saves the `result_metrics` directory a second time within `gtsfm/rtf_vis_tool/src`. To do this, I will need to add some lines to `scene_optimizer.py` and `multi_view_optimizer.py` to save duplicate copies that my React app can access. For my weekly demos on Tuesday, I have just been copy pasting the files into my React source folder, but once I finish merging all my current React code into master, I'll make the necessary edits to the 2 python files. Here's what I need to do next to get a fully functioning React app on our master branch:

- finish merging React PR #2 (containing Components related to the point cloud)
- create and merge React PR #3 (containing Components related to frontend/multiview optimizer metrics)
- create and merge a PR with edits to `scene_optimizer.py` and `multi_view_optimizer.py`. (at this point, the React app should be fully functional)

johnwlambert commented on rtf_vis_tool/src/stylesheets/LandingPageGraph.css:
```diff
@@ -0,0 +1,159 @@
+.navbar {
+    height: 5%;
+    width: 100%;
+    background-color: #D2B48C;
```
Thanks for all the updates, Adi.

One note here -- "tan" is not my favorite background color -- could you use one of the softer Tango color tones instead? https://en.wikipedia.org/wiki/Tango_Desktop_Project (as background)

johnwlambert commented on rtf_vis_tool/src/stylesheets/Div_Node.css:
```diff
@@ -0,0 +1,20 @@
+/*  Gives the DivNode a black curved border with some padding to center the text within it. 
```
do you mind defining what a DivNode is here?

johnwlambert commented on rtf_vis_tool/src/Components/gtsfm_node_list.js:
```diff
@@ -2,48 +2,48 @@
+                        'Post-Processor',
+                        'Putative Correspondence Indices',
+                        'Putative Matcher',
+                        'relative Rs: i2_r_i1',
```
should be an upper case R here, `i2Ri1`

johnwlambert commented on rtf_vis_tool/src/Components/gtsfm_node_list.js:
```diff
@@ -2,48 +2,48 @@
+                        'Putative Matcher',
+                        'relative Rs: i2_r_i1',
+                        'relative Rs (2): i2_R_i1',    
+                        'relative Ts: i2_t_i1',  
```
`i2ti1` is our current notation -- that's my bad for having the wrong thing in the slide picture. I'll update it on my end too

johnwlambert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -76,15 +75,15 @@ const DivGraph = (props) => {
             {/* {showFS && <FrontendSummary json={fs_json} toggleFS={toggleFrontEndSummaryDisplay}/>} */}
             {/* {showMVO && <MVOSummary json={mvo_json} toggleMVO={toggleMVOMetrics}/>} */}
-            {showDA_PC && <Data_Association_PC da_json={da_json} rotated_json={rotated_da_json} toggleDA_PC={toggleDataAssoc_PointCloud}/>}
+            {showDA_PC && <Bundle_Adj_PC da_json={da_json} rotated_json={rotated_da_json} toggleDA_PC={toggleDataAssoc_PointCloud}/>}
```
is the code above meant to be commented out?
```
            {/* Render popups only when the respective node is clicked*/}
            {/* {showFS && <FrontendSummary json={fs_json} toggleFS={toggleFrontEndSummaryDisplay}/>} */}
            {/* {showMVO && <MVOSummary json={mvo_json} toggleMVO={toggleMVOMetrics}/>} */}
```

adisingh50 replied:
yes. I have just deleted a few of those commented out components entirely for the PR. I'll add them back in the next one.

johnwlambert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -1,25 +1,24 @@
-import data_association_json from '.././result_metrics/data_association_metrics.json';
-import frontend_summary_json from '.././result_metrics/frontend_summary.json';
-import multiview_optimizer_json from '.././result_metrics/multiview_optimizer_metrics.json';
+import data_association_json from '../result_metrics/data_association_metrics.json';
```
is there a way to just have the user run a `setup.sh` script that generates a symbolic link, so that only one copy of the files is needed?

```bash
ln -s result_matrics rtf_vis_tool/result_metrics
```
or something like that?

adisingh50 replied:
After looking more into Linux symbolic links, this looks like a great way to get around saving two copies of the file. I'll add a `setup.sh` file within the `rtf_vis_tool` directory that creates symbolic links to the `results` and `result_metrics` directories.

adisingh50 replied:
I've gotten a Linux symlink to connect the directories successfully, but React is still having issues with it. Messaged you privately about it in more detial.

johnwlambert commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -20,41 +25,41 @@ const getGridLines = () => {
 
           <Line 
-            points={[[0,-10,0], [0,10,0]]} 
+            points={[[0,lowerGridBound,0], [0,upperGridBound,0]]} 
```
it looks like we are naming "y" as "z". Could you add a comment with a link to the React documentation explaining why we do this?

adisingh50 replied:
yup, I'll add some more info

johnwlambert commented on rtf_vis_tool/src/App.js:
```diff
@@ -3,13 +3,13 @@ import {extend} from "react-three-fiber";
 import './App.css';
 
-import DivGraph from './Components/DivGraph';
+import LandingPageGraph from './Components/LandingPageGraph';
```
can we add a unit test?

https://testing-library.com/docs/react-testing-library/intro/
https://itnext.io/how-to-unit-test-in-react-72e911e2b8d

adisingh50 replied:
I've added some unit tests for smaller, helper components that do not involve React Three Fiber. This is because the React unit testing framework, Enzyme, does not provide native support for third party packages like React Three Fiber. Thus, it's incompatible with components with render 3D-related components.

johnwlambert commented:
Adi, have you tried a linter like https://medium.com/@RossWhitehouse/setting-up-eslint-in-react-c20015ef35f7?

johnwlambert commented on rtf_vis_tool/src/Components/Bundle_Adj_PC.js:
```diff
@@ -0,0 +1,116 @@
+    //render points3D.txt from COLMAP ba_output directory
+    useEffect(() => {
+        //fetch the specific file from the public directory
+        fetch('results/ba_output/points3D.txt')
```
where is this relative to?

adisingh50 replied:
this is relative to `rtf_vis_tool/public` directory. For reading local text files, I have to use the `fetch()` function, which can only read files that are within `rtf_vis_tool/public`. But for JSON files, I can nicely import them (which is what I do with `data_association_metrics.json`)

johnwlambert commented on rtf_vis_tool/src/Components/Bundle_Adj_PC.js:
```diff
@@ -0,0 +1,116 @@
+            })
+    }, []);
+
+    const updatePointSizes = (radius) => {
```
need a comment/docstring here

johnwlambert commented on rtf_vis_tool/src/Components/DivNode.js:
```diff
@@ -0,0 +1,27 @@
+import React from "react";
+import '.././stylesheets/Div_Node.css';
+
+//Basic Div Component repeatedly rendered in DivGraph.js
```
was DivGraph renamed?

adisingh50 replied:
yup. forgot to update the comment. I've updated it now.

johnwlambert commented on rtf_vis_tool/src/Components/gtsfm_edge_list.js:
```diff
@@ -0,0 +1,58 @@
+//List of all Edges in GTSFM Visualization Graph
+//imported in DivGraph.js
```
was DivGraph renamed?

adisingh50 replied:
yup. forgot to update the comment. I've updated it now.

adisingh50 commented:
John, the default React project actually comes pre installed with a linter. The linter is defined by the `@testing-library/react` dependency within `package.json`. Definitely helpful for removing unneccessary imports.

johnwlambert commented:
> John, the default React project actually comes pre installed with a linter. The linter is defined by the `@testing-library/react` dependency within `package.json`. Definitely helpful for removing unneccessary imports.

cool, thanks

dellaert commented on rtf_vis_tool/README.md:
```diff
@@ -17,7 +17,13 @@ python gtsfm/runner/run_scene_optimizer.py
 This will populate the `result_metrics` directory
 
-4. Now, run the web application:
+4. Run the React Unit Tests:
```
Move this to (2) and make optional

dellaert commented on rtf_vis_tool/README.md:
````diff
@@ -17,7 +17,13 @@ python gtsfm/runner/run_scene_optimizer.py
+```
+Currently, the unit tests are written for smaller, helper components that do not involve React Three Fiber. This is because the React unit testing framework, Enzyme, does not provide native support for third party packages like React Three Fiber. Thus, it's incompatible with components with render 3D-related components.
+
+5. Now, run the web application:
````
move this to 3.

dellaert commented on rtf_vis_tool/README.md:
```diff
@@ -29,7 +35,7 @@ Note: running the application for the first time will auto generate a `/node_mod
 - `public`: contains index.html file 
 - `src`
-    - `Components`: contains all React components used for rendering graph nodes, summaries, point cloud viewers, etc...
+    - `Components`: contains all React components used for rendering graph nodes, summaries, point cloud viewers, etc. Each file within this folder returns a format JSX, which is essentially a modified version of HTML that is meant to be readable by React code
```
emphasize that index.html contains the initial view that will be displayed, and that it will start App.js, which is the top-level component. Only keep App.js and rename it to `GtsfmVisualizer.js`

adisingh50 replied:
To make things simpler, I've just removed `App.js` all together (since it was never really necessary). Now the hierarchy is 
`index.html` -> `index.js` -> `LandingPageGraph.js`

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+import React, {useEffect, useState} from "react";
```
Add comment block like in python to *all* js files.
/*
Copyright

file

author
*/

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+import Xarrow from "react-xarrows";
+import '../stylesheets/LandingPageGraph.css'
+
+//Loading Components
```
Use formatter in code, configure Code to auto-format. Use Google style if it can deal with react.

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+import Bundle_Adj_PC from './Bundle_Adj_PC';
+import DivNode from './DivNode';
+import EdgeList from './gtsfm_edge_list.js';
+// import FrontendSummary from './FrontendSummary';         - add this in part 3
```
delete

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+import frontend_summary_json from '../result_metrics/frontend_summary.json';
+import multiview_optimizer_json from '../result_metrics/multiview_optimizer_metrics.json';
+
+//Landing Page Component to display GTSFM graph (what user first sees)
```
// Landing Page Component to display GTSFM graph (what user first sees).

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+
+//Landing Page Component to display GTSFM graph (what user first sees)
+const LandingPageGraph = (props) => {
+    const [arrowList, setArrowList] = useState([]);
```
Maybe add a comment on top each block

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+    const [showRSS, setShowRSS] = useState(false);
+    const [rotated_da_json, setRotatedDAJSON] = useState(null);
+
+    //render all directed edges on the graph
```
`// Render all directed edges on the graph.`

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+import React, {useEffect, useState} from "react";
+import Xarrow from "react-xarrows";
```
Add `// Used to render edges.`

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+    const [rotated_da_json, setRotatedDAJSON] = useState(null);
+
+    //render all directed edges on the graph
+    //save all the json resulting metrics in separate React variables
```
Move these two comments to the blocks

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,160 @@
+        setRotatedDAJSON(data_association_json.rotated_points_3d);
+    }, [])
+
+    //Function to defines the percent offset each div node has from the top and left side of the screen
```
`// Function to define...`

dellaert commented on rtf_vis_tool/src/Components/gtsfm_edge_list.js:
```diff
@@ -0,0 +1,58 @@
+//List of all Edges in GTSFM Visualization Graph
+//imported in LandingPageGraph.js
+
+const gtsfm_edge_list = [['Scene Image Directories', 'Scene Looper'], 
```
- Add a map from string to grey nodes, add drawing information and anything else
- Add a map from string to blue nodes, add drawing information and anything else
- the edges below are fine

adisingh50 replied:
done

dellaert commented on rtf_vis_tool/src/Components/gtsfm_edge_list.js:
```diff
@@ -0,0 +1,58 @@
+//List of all Edges in GTSFM Visualization Graph
```
rename to gtsfm_graph

// Saves the bipattite graph (U,V,E), where U are the (gray) products, and (V) are the (blue) processes.

adisingh50 replied:
I ended up having to store the graph in 3 separate JS files, because in any file, I can only export one variable. So, instead, I have created a directory called `gtsfm_graph`, and inside that directory, I have:

- `gray_nodes.js`: stores all the gray nodes
- `blue_nodes.js`: stores all the blue nodes
- `edge_list.js`: stores all the directed edges

dellaert commented on rtf_vis_tool/src/Components/Bundle_Adj_PC.js:
```diff
@@ -0,0 +1,125 @@
+            .then(function(response){
+                return response.text();
+            })
+            .then(function(data){
```
Consider defining function somewhere else rather than inline, with a descriptive name and it’s own comment

dellaert commented on rtf_vis_tool/src/Components/Bundle_Adj_PC.js:
```diff
@@ -0,0 +1,125 @@
+    /* Function that updates the radius of all points within a point cloud.
+       Called everytime the react slider input is interacted with.
+    */
+    const updatePointSizes = (radius) => {
```
Is this the most modern style of declaring functions?

adisingh50 replied:
I believe so. Defining functions through this concise arrow syntax is the preferred method.

dellaert commented on rtf_vis_tool/src/Components/Bundle_Adj_PC.js:
```diff
@@ -0,0 +1,125 @@
+                Toggle Coordinate Grid
+            </button>
+
+            <div className="point_size_slider">
```
Consider making into a separate component. In general , no function/component should be longer than 4 lines, give or take 5 ;-)

dellaert commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -0,0 +1,80 @@
+    return finalLineGridSet;
+}
+
+const CoordinateGrid = () => { 
```
Docstring?

adisingh50 replied:
Since `CoordinateGrid` is the main component of the file, I have its docstring at the top of the file. I have done this for all components.

dellaert commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -0,0 +1,80 @@
+const upperGridBound = 10;    // Upper grid bound for coordinate axes.
+const htmlSize = 50;          // Size of the x,y,z axes labels.
+
+// Function that creates the grid lines along with xy plane.
```
Of course it’s a function. Comments should be command-like, not descriptive:
```
/// Create grid lines along with xy plane.
```

dellaert commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -0,0 +1,80 @@
+// Third-Party Package Imports.
+import {Line, Html} from "drei";  // Used to render axes lines and text.
+
+const lowerGridBound = -10;   // Lower grid bound for coordinate axes.
```
Consider making these file-level constants default arguments to the function below.

dellaert commented on rtf_vis_tool/src/Components/DivNode.js:
```diff
@@ -0,0 +1,34 @@
+/* Basic Div Component repeatedly rendered which is rendered within LandingPageGraph.js. Called a 
```
Then call it Node?

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,172 @@
+        var grayNodes = GrayNodes;
+        var grayNodes_formatted = [];
+        for (var j = 0; j < grayNodes.length; j++) {
+            const nodeText = grayNodes[j].text;
```
Make a separate component GrayNode, pass it grayNodes[j]

dellaert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,172 @@
+        var blueNodes = BlueNodes;
+        var blueNodes_formatted = [];
+        for (var k = 0; k < blueNodes.length; k++) {
+            const nodeText = blueNodes[k].text;
```
<BlueNode node=blueNodes[k]>

adisingh50 replied:
same as above done for `BlueNode.js`

dellaert commented on rtf_vis_tool/src/Components/gtsfm_graph/blue_nodes.js:
```diff
@@ -0,0 +1,99 @@
+/* List of all Blue Nodes in GTSFM Visualization Graph. Imported in LandingPageGraph.js.
```
nice!

dellaert reviewed (approved):
Very cool! Approving modulo comments, but I trust you :-)

johnwlambert commented on rtf_vis_tool/src/tests/DivNode.test.js:
```diff
@@ -0,0 +1,58 @@
+import Adapter from 'enzyme-adapter-react-16'
+
+// Local Imports.
+import DivNode from '../Components/DivNode';
```
is DivNode outdated now?

Is there any way to run the unit tests inside our CI? what's the one-liner?

adisingh50 replied:
Yes, DivNode is outdated now. I have updated the tests with the new, correct names. To run the unit tests, just run `npm test` within the `gtsfm/rtf_vis_tool` directory.

johnwlambert commented on rtf_vis_tool/src/Components/Node.js:
```diff
@@ -0,0 +1,33 @@
+// Local Imports.
+import '.././stylesheets/Node.css';
+
+const Node = (props) => {
```
@dellaert maybe this is too minor, but maybe we should use `GtsfmNode` instead of `Node`? Since "Node" already has a strong meaning in Javascript?

adisingh50 replied:
I agree. It may be a bit misleading to name a node `Node.js` since it already has a strong meaning in the JS world. Will rename it to `GtsfmNode`

johnwlambert commented:
Adi, feel free to add yourself to the author list in the Readme, as well

johnwlambert commented on rtf_vis_tool/src/Components/Bundle_Adj_PC.js:
```diff
@@ -0,0 +1,114 @@
+    const updatePointSizes = (radius) => {
+        var finalPointsJSX = [];
+        for (var i = 0; i < pointCloudRaw.length; i += 1) {
+            var pointArr = pointCloudRaw[i];
```
is `pointCloudRaw` a global variable?

johnwlambert replied:
maybe an inline comment that `pointCloudRaw` is (N x 6) array, with first 3 columns as (x,y,z) and last 3 columns as (R,G,B) would be helpful

adisingh50 replied:
so `pointCloudRaw` is a state variable, which is just a React term for saying its only accessible within the `Bundle_Adj_PC` component. So I'd think of it more as a local variable. Anything that is defined like `const [pointCloudRaw, setPointCloudRaw] = useState([]);` is considered a react state which is locally defined to the component

johnwlambert commented on rtf_vis_tool/src/Components/CoordinateGrid.js:
```diff
@@ -0,0 +1,79 @@
+// Third-Party Package Imports.
+import {Line, Html} from "drei";  // Used to render axes lines and text.
+
+// Create grid lines along the xy plane.
```
what are the grid lines for? maybe that can be mentioned in the comment? to help orient the user?

johnwlambert commented on rtf_vis_tool/src/Components/LandingPageGraph.js:
```diff
@@ -0,0 +1,141 @@
+import Node from './Node';
+import '../stylesheets/LandingPageGraph.css'
+
+const LandingPageGraph = (props) => {
```
docstring missing here?

johnwlambert commented on rtf_vis_tool/src/Components/OrbitControlsComponent.js:
```diff
@@ -0,0 +1,28 @@
+// Third-Party Package Imports.
+import {useThree} from "react-three-fiber"; // Used to define a three.js camera object.
+
+const OrbitControlsComponent = () => {
```
missing docstring?

johnwlambert commented on rtf_vis_tool/src/Components/PointMesh.js:
```diff
@@ -0,0 +1,16 @@
+*/
+import React from "react";
+
+const PointMesh = (props) => {
```
missing docstring?

johnwlambert commented on rtf_vis_tool/src/Components/PointSizeSlider.js:
```diff
@@ -0,0 +1,32 @@
+// Local Imports.
+import '.././stylesheets/PointSizeSlider.css';
+
+const PointSizeSlider = (props) => {
```
missing docstring?

johnwlambert reviewed (approved):
@adisingh50 I left a few minor comments, I'm happy to merge it after those few small changes

adisingh50 commented:
Added the docstrings for each component and changed all functions to look like this:
![image](https://user-images.githubusercontent.com/40643713/113963411-8809ec00-97f7-11eb-98e5-4a745c2a2bdb.png)

instead of this:
![image](https://user-images.githubusercontent.com/40643713/113963434-922bea80-97f7-11eb-8397-51262b07bebf.png)

johnwlambert commented on rtf_vis_tool/src/Components/Bundle_Adj_PC.js:
```diff
@@ -0,0 +1,133 @@
+    const pointSizeArr = [pointRadius]; 
+
+    // Render points3D.txt from COLMAP ba_input directory.
+    useEffect(() => {
```
is this the previous function style?
